metaflow/plugins/argo/event_decorators.py

```diff
@@ -0,0 +1,242 @@
+
+Examples
+--------
+@{0}(flow="MyFirstFlow")
+@{0}(flows=["MyFirstFlow", "MySecondFlow"])
+"""
+
+BAD_INPUTS_MSG = """The {0} attribute must be a list of {1} names.
+
+Examples
+--------
+@{2}({0}=["MyOtherFlow"])
+"""
+
+BAD_AGGREGATE_MAPPING_MSG = """Event field mappings for multiple events or flows must be a dict of
+string keys corresponding to an event or flow name and values of another
+dict of string keys and values mapping flow parameters to event fields.
+
+Examples
+--------
+@trigger_on(flows=["FirstFlow", "SecondFlow"], mappings={"FirstFlow": {"alpha": "alpha_value"}, 
+                                                         "SecondFlow": {"delta": "delta_value"}})
+@trigger_on(events=["first", "second"], mappings={"first": {"alpha": "alpha_value"}, 
+                                                  "second": {"delta": "delta_value"}})                                                         
+"""
+
+BAD_MAPPING_MSG = """Event field mappings for a single event or flow must be a dict of
+string keys and values mapping flow parameters to event fields.
+
+Examples
+--------
+@trigger_on(flow="FirstFlow", mappings={"alpha": "alpha_value", "delta": "delta_value"})
+@trigger_on(event="first", mappings={"alpha": "alpha_value", "delta": "delta_value"})
+"""
+
+
+class BadEventNameException(MetaflowException):
+    headline = "Bad or missing event name detected"
+
+
+class BadEventMappingsException(MetaflowException):
+    headline = "Bad event field mappings detected"
+
+
+class BadTriggerInputsException(MetaflowException):
+    headline = "Bad workflow trigger inputs detected"
+
+
+BAD_AGGREGATE_MAPPING_ERROR = BadEventMappingsException(msg=BAD_AGGREGATE_MAPPING_MSG)
+
+BAD_MAPPING_ERROR = BadEventMappingsException(msg=BAD_MAPPING_MSG)
+
+
+def bad_inputs_error(field_name, field_type, deco_name):
+    return BadTriggerInputsException(
+        msg=BAD_INPUTS_MSG.format(field_name, field_type, deco_name, field_name)
+    )
+
+
+def validate_mappings(mappings, aggregate):
+    if mappings is None:
+        return None
+    if type(mappings) != dict:
+        if aggregate:
+            raise BAD_AGGREGATE_MAPPING_ERROR
+        else:
+            raise BAD_MAPPING_ERROR
+    for k in mappings.keys():
+        if type(k) != str:
+            if aggregate:
+                raise BAD_AGGREGATE_MAPPING_ERROR
+            else:
+                raise BAD_MAPPING_ERROR
+        v = mappings[k]
+        if aggregate:
+            if type(v) != dict:
+                raise BAD_AGGREGATE_MAPPING_ERROR
+            for k1 in v.keys():
+                v1 = v[k1]
+                if type(k1) != str or type(v1) != str:
+                    raise BAD_AGGREGATE_MAPPING_ERROR
+        else:
+            if type(v) != str:
+                raise BAD_MAPPING_ERROR
+    return mappings
+
+
+class TriggerOnDecorator(FlowDecorator):
+
+    name = "trigger_on"
+    defaults = {
+        "flow": None,
```

Comment:
@savingoyal : I don't think we need to have flow and flows and event and events. We can stick to just one and take a list (or just consider a single element to be a list of a single element). IMHO, the API should be fairly straightfoward.

I would also not stick to just "flows" because we can trigger at the task level (task is not user visible but we should be able to allow users to trigger on steps at the very least).

Reply:
@savingoyal : I also suppose this is an "and" type of situation (ie: the trigger only happens if everything specified here matches). Is this the assumption?

Reply:
@savingoyal : these fields also need to be deploy-time evaluated. I don't think this is a huge problem but mentioning it as this is a fairly key feature in some of our workflows.